internal/filecopy/filecopy.go

```diff
@@ -0,0 +1,111 @@
+package filecopy
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/iver-wharf/wharf-cmd/internal/ignorer"
+)
+
+// Copier is an interface that can copy bytes from one reader to a writer.
+type Copier interface {
+	Copy(dst io.Writer, src io.Reader) error
+}
+
+type ioCopier struct{}
+
+func (ioCopier) Copy(dst io.Writer, src io.Reader) error {
+	_, err := io.Copy(dst, src)
+	return err
+}
+
+// IOCopier is a copier implementation that uses the built-in io.Copy()
+// to copy the content in a loop.
+var IOCopier Copier = ioCopier{}
+
+// CopyFile takes a destination and source path and copies the content via the
+// copier implementation.
+func CopyFile(dst, src string, copier Copier) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("open src: %w", err)
+	}
+	defer srcFile.Close()
+
+	srcFileStat, err := srcFile.Stat()
+	if err != nil {
+		return fmt.Errorf("stat src: %w", err)
+	}
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, srcFileStat.Mode())
+	if err != nil {
+		return fmt.Errorf("open dst: %w", err)
+	}
+	defer dstFile.Close()
+
+	return copier.Copy(dstFile, srcFile)
+}
```

Comment:
This function seems like it might be called a lot, in which case changing to non-defer closing might be provide further speed increase. Unsure how much of an overhead `defer` has in Go v1.18, but adding this as a note for later regardless.

Reply:
This is premature optimization, based on a heuristic. Fixing this would cost complexity, for probably little gain.

Better to run profiling toolchains to find the real culprits in the code, and besides the slow part of wharf-cmd right now isn't wharf-cmd but instead the waiting for Kubernetes pods to be started.

Besides, the `defer` keyword has been getting a lot of optimizations throughout the years. In Go 1.18, from my personal testing, I get something around 3.5 ns/op overhead from the `defer` call. This function contains two defers, so on one of our larger internal projects that contains ~3000 code non-gitignored files then these defers adds around 20-24 µs in total overhead.

Closing this as "wont-fix"

Reply:
I know I did this kind of "defer-less dance" before in

https://github.com/iver-wharf/wharf-cmd/blob/8263c5fd11de1423019e44d1c82eab26edbce227/pkg/resultstore/logline.go#L109-L113

But that's before I realized how little overhead the defer actually has. My heuristic I used then was based on very old blog posts where people complained about `defer`s performance in Go 1.1 and such.